// mup/lib/modules/meteor/index.js
'use strict';

require('babel-polyfill');
require('source-map-support/register');

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.help = help;
exports.logs = logs;
exports.setup = setup;
exports.push = push;
exports.envconfig = envconfig;
exports.start = start;
exports.deploy = deploy;
exports.stop = stop;

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _debug = require('debug');

var _debug2 = _interopRequireDefault(_debug);

var _nodemiral = require('nodemiral');

var _nodemiral2 = _interopRequireDefault(_nodemiral);

var _uuid = require('uuid');

var _uuid2 = _interopRequireDefault(_uuid);

var _underscore = require('underscore');

var _ = _interopRequireWildcard(_underscore);

var _utils = require('../utils');

var _build = require('./build.js');

var _build2 = _interopRequireDefault(_build);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var log = (0, _debug2.default)('mup:module:meteor');

function help() /* api */{
  log('exec => mup meteor help');
  console.log('mup meteor', Object.keys(this));
}

function logs(api) {
  log('exec => mup meteor logs');
  var config = api.getConfig().meteor;
  if (!config) {
    console.error('error: no configs found for meteor');
    process.exit(1);
  }

  var args = api.getArgs();
  var sessions = api.getSessions(['meteor']);
  return (0, _utils.getDockerLogs)(config.name, sessions, args);
}

function setup(api) {
  log('exec => mup meteor setup');
  var config = api.getConfig().meteor;
  if (!config) {
    console.error('error: no configs found for meteor');
    process.exit(1);
  }

  var list = _nodemiral2.default.taskList('Setup Meteor');

  list.executeScript('Setup Environment', {
    script: _path2.default.resolve(__dirname, 'assets/meteor-setup.sh'),
    vars: {
      name: config.name
    }
  });

  if (config.ssl) {
    var basePath = api.getBasePath();

    if (config.ssl.upload !== false) {
      list.copy('Copying SSL Certificate Bundle', {
        src: _path2.default.resolve(basePath, config.ssl.crt),
        dest: '/opt/' + config.name + '/config/bundle.crt'
      });

      list.copy('Copying SSL Private Key', {
        src: _path2.default.resolve(basePath, config.ssl.key),
        dest: '/opt/' + config.name + '/config/private.key'
      });
    }

    list.executeScript('Verifying SSL Configurations', {
      script: _path2.default.resolve(__dirname, 'assets/verify-ssl-config.sh'),
      vars: {
        name: config.name
      }
    });
  }

  var sessions = api.getSessions(['meteor']);

  return (0, _utils.runTaskList)(list, sessions);
}

function push(api) {
  log('exec => mup meteor push');
  var config = api.getConfig().meteor;
  if (!config) {
    console.error('error: no configs found for meteor');
    process.exit(1);
  }
  if (!config.docker) {
    if (config.dockerImage) {
      config.docker = { image: config.dockerImage };
      delete config.dockerImage;
    } else {
      config.docker = { image: 'kadirahq/meteord' };
    }
  }
  if (config.dockerImageFrontendServer) {
    config.docker.imageFrontendServer = config.dockerImageFrontendServer;
  }
  if (!config.docker.imageFrontendServer) {
    config.docker.imageFrontendServer = 'meteorhacks/mup-frontend-server';
  }

  var buildOptions = config.buildOptions || {};
  buildOptions.buildLocation = buildOptions.buildLocation || _path2.default.resolve('/tmp', _uuid2.default.v4());

  console.log('Building App Bundle Locally');

  var bundlePath = _path2.default.resolve(buildOptions.buildLocation, 'bundle.tar.gz');
  var appPath = _path2.default.resolve(api.getBasePath(), config.path);

  return (0, _build2.default)(appPath, buildOptions).then(function () {
    config.log = config.log || {
      opts: {
        'max-size': '100m',
        'max-file': 10
      }
    };
    var list = _nodemiral2.default.taskList('Pushing Meteor');

    list.copy('Pushing Meteor App Bundle to The Server', {
      src: bundlePath,
      dest: '/opt/' + config.name + '/tmp/bundle.tar.gz',
      progressBar: config.enableUploadProgressBar
    });

    list.copy('Pushing the Startup Script', {
      src: _path2.default.resolve(__dirname, 'assets/templates/start.sh'),
      dest: '/opt/' + config.name + '/config/start.sh',
      vars: {
        appName: config.name,
        useLocalMongo: api.getConfig().mongo ? 1 : 0,
        port: config.env.PORT || 80,
        sslConfig: config.ssl,
        logConfig: config.log,
        volumes: config.volumes,
        docker: config.docker
      }
    });

    var sessions = api.getSessions(['meteor']);
    return (0, _utils.runTaskList)(list, sessions, { series: true });
  });
}

function envconfig(api) {
  log('exec => mup meteor envconfig');
  var config = api.getConfig().meteor;
  if (!config) {
    console.error('error: no configs found for meteor');
    process.exit(1);
  }

  var list = _nodemiral2.default.taskList('Configuring  Meteor Environment Variables');

  var env = _.clone(config.env);
  env.METEOR_SETTINGS = JSON.stringify(api.getSettings());
  // sending PORT to the docker container is useless.
  // It'll run on PORT 80 and we can't override it
  // Changing the port is done via the start.sh script
  delete env.PORT;

  list.copy('Sending Environment Variables', {
    src: _path2.default.resolve(__dirname, 'assets/templates/env.list'),
    dest: '/opt/' + config.name + '/config/env.list',
    vars: {
      env: env || {},
      appName: config.name
    }
  });
  var sessions = api.getSessions(['meteor']);
  return (0, _utils.runTaskList)(list, sessions, { series: true });
}

function start(api) {
  log('exec => mup meteor start');
  var config = api.getConfig().meteor;
  if (!config) {
    console.error('error: no configs found for meteor');
    process.exit(1);
  }

  var list = _nodemiral2.default.taskList('Start Meteor');

  list.executeScript('Start Meteor', {
    script: _path2.default.resolve(__dirname, 'assets/meteor-start.sh'),
    vars: {
      appName: config.name
    }
  });

  list.executeScript('Verifying Deployment', {
    script: _path2.default.resolve(__dirname, 'assets/meteor-deploy-check.sh'),
    vars: {
      deployCheckWaitTime: config.deployCheckWaitTime || 60,
      appName: config.name,
      port: config.env.PORT || 80
    }
  });

  var sessions = api.getSessions(['meteor']);
  return (0, _utils.runTaskList)(list, sessions, { series: true });
}

function deploy(api) {
  log('exec => mup meteor deploy');
  var config = api.getConfig().meteor;
  if (!config) {
    console.error('error: no configs found for meteor');
    process.exit(1);
  }

  return push(api).then(function () {
    return envconfig(api);
  }).then(function () {
    return start(api);
  });
}

function stop(api) {
  log('exec => mup meteor stop');
  var config = api.getConfig().meteor;
  if (!config) {
    console.error('error: no configs found for meteor');
    process.exit(1);
  }

  var list = _nodemiral2.default.taskList('Stop Meteor');

  list.executeScript('Stop Meteor', {
    script: _path2.default.resolve(__dirname, 'assets/meteor-stop.sh'),
    vars: {
      appName: config.name
    }
  });

  var sessions = api.getSessions(['meteor']);
  return (0, _utils.runTaskList)(list, sessions);
}
