// mup/lib/modules/utils.js
'use strict';

require('babel-polyfill');
require('source-map-support/register');

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.runTaskList = runTaskList;
exports.getDockerLogs = getDockerLogs;
exports.runSSHCommand = runSSHCommand;
exports.countOccurences = countOccurences;

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

var _underscore = require('underscore');

var _ = _interopRequireWildcard(_underscore);

var _bluebird = require('bluebird');

var _ssh = require('ssh2');

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function runTaskList(list, sessions, opts) {
  return new Promise(function (resolve, reject) {
    list.run(sessions, opts, function (summaryMap) {
      for (var host in summaryMap) {
        var summary = summaryMap[host];
        if (summary.error) {
          reject(summary.error);
          return;
        }
      }

      resolve();
    });
  });
}

function getDockerLogs(name, sessions, args) {
  var command = 'sudo docker ' + args.join(' ') + ' ' + name;

  var promises = _.map(sessions, function (session) {
    var host = '[' + session._host + ']';
    var options = {
      onStdout: function onStdout(data) {
        process.stdout.write(host + data);
      },
      onStderr: function onStderr(data) {
        process.stdout.write(host + data);
      }
    };
    return (0, _bluebird.promisify)(session.execute.bind(session))(command, options);
  });
  return Promise.all(promises);
}

// Maybe we should create a new npm package
// for this one. Something like 'sshelljs'.
function runSSHCommand(info, command) {
  return new Promise(function (resolve, reject) {
    var conn = new _ssh.Client();

    // TODO better if we can extract SSH agent info from original session
    var sshAgent = process.env.SSH_AUTH_SOCK;
    var ssh = {
      host: info.host,
      port: info.opts && info.opts.port || 22,
      username: info.username
    };

    if (info.pem) {
      ssh.privateKey = _fs2.default.readFileSync(_path2.default.resolve(info.pem), 'utf8');
    } else if (info.password) {
      ssh.password = info.password;
    } else if (sshAgent && _fs2.default.existsSync(sshAgent)) {
      ssh.agent = sshAgent;
    }
    conn.connect(ssh);

    conn.once('error', function (err) {
      if (err) {
        reject(err);
      }
    });

    // TODO handle error events
    conn.once('ready', function () {
      conn.exec(command, function (err, stream) {
        if (err) {
          conn.end();
          reject(err);
          return;
        }

        var output = '';

        stream.on('data', function (data) {
          output += data;
        });

        stream.once('close', function (code) {
          conn.end();
          resolve({ code: code, output: output });
        });
      });
    });
  });
}

function countOccurences(needle, haystack) {
  var regex = new RegExp(needle, 'g');
  var match = haystack.match(regex) || [];
  return match.length;
}
