// mup/lib/mup-api.js
'use strict';

require('babel-polyfill');
require('source-map-support/register');

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _nodemiral = require('nodemiral');

var _nodemiral2 = _interopRequireDefault(_nodemiral);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var MupAPI = function () {
  function MupAPI(base, args, configPath, settingsPath) {
    _classCallCheck(this, MupAPI);

    this.base = base;
    this.args = args;
    this.config = null;
    this.settings = null;
    this.sessions = null;
    this.configPath = configPath;
    this.settingsPath = settingsPath;
  }

  _createClass(MupAPI, [{
    key: 'getArgs',
    value: function getArgs() {
      return this.args;
    }
  }, {
    key: 'getBasePath',
    value: function getBasePath() {
      return this.base;
    }
  }, {
    key: 'getConfig',
    value: function getConfig() {
      if (!this.config) {
        var filePath = void 0;
        if (this.configPath) {
          filePath = _path2.default.resolve(this.configPath);
          this.base = _path2.default.dirname(this.configPath);
        } else {
          filePath = _path2.default.join(this.base, 'mup.js');
        }
        try {
          this.config = require(filePath);
        } catch (e) {
          if (e.code == 'MODULE_NOT_FOUND') {
            console.error('\'mup.js\' file not found. Run \'mup init\' first.');
          } else {
            console.error(e);
          }
          process.exit(1);
        }
      }

      return this.config;
    }
  }, {
    key: 'getSettings',
    value: function getSettings() {
      if (!this.settings) {
        var filePath = void 0;
        if (this.settingsPath) {
          filePath = _path2.default.resolve(this.settingsPath);
        } else {
          filePath = _path2.default.join(this.base, 'settings.json');
        }
        this.settings = require(filePath);
      }

      return this.settings;
    }
  }, {
    key: 'getSessions',
    value: function getSessions() {
      var modules = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : [];

      var sessions = this._pickSessions(modules);
      return Object.keys(sessions).map(function (name) {
        return sessions[name];
      });
    }
  }, {
    key: 'withSessions',
    value: function withSessions() {
      var modules = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : [];

      var api = Object.create(this);
      api.sessions = this._pickSessions(modules);
      return api;
    }
  }, {
    key: '_pickSessions',
    value: function _pickSessions() {
      var _this = this;

      var modules = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : [];

      if (!this.sessions) {
        this._loadSessions();
      }

      var sessions = {};

      modules.forEach(function (moduleName) {
        var moduleConfig = _this.config[moduleName];
        if (!moduleConfig) {
          return;
        }

        for (var name in moduleConfig.servers) {
          if (!moduleConfig.servers.hasOwnProperty(name)) {
            continue;
          }

          if (_this.sessions[name]) {
            sessions[name] = _this.sessions[name];
          }
        }
      });

      return sessions;
    }
  }, {
    key: '_loadSessions',
    value: function _loadSessions() {
      var config = this.getConfig();
      this.sessions = {};

      // `mup.servers` contains login information for servers
      // Use this information to create nodemiral sessions.
      for (var name in config.servers) {
        if (!config.servers.hasOwnProperty(name)) {
          continue;
        }

        var info = config.servers[name];
        var auth = { username: info.username };
        var opts = { ssh: {} };

        var sshAgent = process.env.SSH_AUTH_SOCK;

        if (info.opts) {
          opts.ssh = info.opts;
        }

        if (info.pem) {
          auth.pem = _fs2.default.readFileSync(_path2.default.resolve(info.pem), 'utf8');
        } else if (info.password) {
          auth.password = info.password;
        } else if (sshAgent && _fs2.default.existsSync(sshAgent)) {
          opts.ssh.agent = sshAgent;
        } else {
          console.error('error: server %s doesn\'t have password, ssh-agent or pem', name);
          process.exit(1);
        }

        var session = _nodemiral2.default.session(info.host, auth, opts);
        this.sessions[name] = session;
      }
    }
  }]);

  return MupAPI;
}();

exports.default = MupAPI;
