// mup/lib/modules/default/index.js
'use strict';

require('babel-polyfill');
require('source-map-support/register');

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.deploy = deploy;
exports.help = help;
exports.init = init;
exports.logs = logs;
exports.reconfig = reconfig;
exports.restart = restart;
exports.setup = setup;
exports.start = start;
exports.stop = stop;

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _debug = require('debug');

var _debug2 = _interopRequireDefault(_debug);

var _shelljs = require('shelljs');

var _shelljs2 = _interopRequireDefault(_shelljs);

var _meteor = require('../meteor/');

var meteor = _interopRequireWildcard(_meteor);

var _mongo = require('../mongo/');

var mongo = _interopRequireWildcard(_mongo);

var _docker = require('../docker/');

var docker = _interopRequireWildcard(_docker);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var log = (0, _debug2.default)('mup:module:default');

_shelljs2.default.config.silent = true;

function deploy(api) {
  log('exec => mup deploy');
  return meteor.deploy(api);
}

function help(api) {
  log('exec => mup help');
}
function init() /* api */{
  log('exec => mup init');

  // TODO check if mup.js or settings.json files exists
  var mupJs = _path2.default.resolve(__dirname, 'template/mup.js.sample');
  var settinsJson = _path2.default.resolve(__dirname, 'template/settings.json');
  var mupJsDst = _path2.default.resolve(process.cwd(), 'mup.js');
  var settingsJsonDst = _path2.default.resolve(process.cwd(), 'settings.json');

  _shelljs2.default.cp(mupJs, mupJsDst);
  _shelljs2.default.cp(settinsJson, settingsJsonDst);
}

function logs(api) {
  log('exec => mup logs');
  return meteor.logs(api);
}

function reconfig(api) {
  log('exec => mup reconfig');
  return meteor.envconfig(api).then(function () {
    return meteor.start(api);
  });
}

function restart(api) {
  log('exec => mup restart');
  return meteor.stop(api).then(function () {
    return meteor.start(api);
  });
}

function setup(api) {
  log('exec => mup setup');
  var config = api.getConfig();
  return docker.setup(api).then(function () {
    if (config.mongo) {
      return Promise.all([meteor.setup(api), mongo.setup(api)]).then(function () {
        return mongo.start(api);
      });
    }
    return meteor.setup(api);
  });
}

function start(api) {
  log('exec => mup start');
  return meteor.start(api);
}

function stop(api) {
  log('exec => mup stop');
  return meteor.stop(api);
}
