// mup/lib/modules/mongo/index.js
'use strict';

require('babel-polyfill');
require('source-map-support/register');

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.dump = dump;
exports.help = help;
exports.logs = logs;
exports.setup = setup;
exports.start = start;
exports.stop = stop;

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _debug = require('debug');

var _debug2 = _interopRequireDefault(_debug);

var _nodemiral = require('nodemiral');

var _nodemiral2 = _interopRequireDefault(_nodemiral);

var _utils = require('../utils');

var _docker = require('../docker/');

var docker = _interopRequireWildcard(_docker);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var log = (0, _debug2.default)('mup:module:mongo');

function dump() /* api */{
  log('exec => mup mongo dump');
}

function help() /* api */{
  log('exec => mup mongo help');
}

function logs(api) {
  log('exec => mup mongo logs');

  var args = api.getArgs();
  var sessions = api.getSessions(['mongo']);
  return (0, _utils.getDockerLogs)('mongodb', sessions, args);
}

function setup(api) {
  log('exec => mup mongo setup');

  var mongoSessions = api.getSessions(['mongo']);
  var meteorSessions = api.getSessions(['meteor']);

  if (meteorSessions.length !== 1) {
    console.log('To use mup inbuilt mongodb setup, you should have only one meteor app server. To have more app servers, use an external mongodb setup');
    return;
  } else if (mongoSessions[0]._host !== meteorSessions[0]._host) {
    console.log('To use mup inbuilt mongodb setup, you should have both meteor app and mongodb on the same server');
    return;
  }

  var list = _nodemiral2.default.taskList('Setup Mongo');

  list.executeScript('setup environment', {
    script: _path2.default.resolve(__dirname, 'assets/mongo-setup.sh')
  });

  list.copy('copying mongodb.conf', {
    src: _path2.default.resolve(__dirname, 'assets/mongodb.conf'),
    dest: '/opt/mongodb/mongodb.conf'
  });

  var sessions = api.getSessions(['mongo']);

  return (0, _utils.runTaskList)(list, sessions);
}

function start(api) {
  log('exec => mup mongo start');

  var mongoSessions = api.getSessions(['mongo']);
  var meteorSessions = api.getSessions(['meteor']);

  if (meteorSessions.length !== 1 || mongoSessions[0]._host !== meteorSessions[0]._host) {
    log('Skipping mongodb start. Incompatible config');
    return;
  }

  var list = _nodemiral2.default.taskList('Start Mongo');

  list.executeScript('start mongo', {
    script: _path2.default.resolve(__dirname, 'assets/mongo-start.sh')
  });

  var sessions = api.getSessions(['mongo']);
  return (0, _utils.runTaskList)(list, sessions);
}

function stop(api) {
  log('exec => mup mongo stop');
  var list = _nodemiral2.default.taskList('Stop Mongo');

  list.executeScript('stop mongo', {
    script: _path2.default.resolve(__dirname, 'assets/mongo-stop.sh')
  });

  var sessions = api.getSessions(['mongo']);
  return (0, _utils.runTaskList)(list, sessions);
}
